packages/block-editor/src/components/provider/index.js

```diff
@@ -7,31 +7,40 @@ import { withDispatch, withSelect } from '@wordpress/data';
 import { compose } from '@wordpress/compose';
 
 class BlockEditorProvider extends Component {
-	constructor( props ) {
-		super( ...arguments );
-		props.updateEditorSettings( props.settings );
-		props.initBlocks( props.value );
-		this.persistedValue = props.select( 'core/block-editor' ).getBlocks();
+	componentDidMount() {
+		this.props.updateEditorSettings( this.props.settings );
+		this.props.resetBlocks( this.props.value );
+
+		this.isSyncingBlockValue = true;
 	}
 
 	componentDidUpdate( prevProps ) {
-		if ( this.props.settings !== prevProps.settings ) {
-			this.props.updateEditorSettings( this.props.settings );
+		const {
+			settings,
+			updateEditorSettings,
+			value,
+			resetBlocks,
+			blocks,
+			onChange,
+		} = this.props;
+
+		if ( settings !== prevProps.settings ) {
+			updateEditorSettings( settings );
 		}
 
-		if (
-			this.props.blocks !== prevProps.blocks &&
-			this.props.blocks !== this.persistedValue
-		) {
-			this.persistedValue = this.props.blocks;
-			this.props.onChange( this.props.blocks );
+		if ( this.isSyncingBlockValue ) {
+			this.isSyncingBlockValue = false;
+		} else if ( blocks !== prevProps.blocks ) {
+			onChange( blocks );
+			this.isSyncingBlockValue = true;
```

Comment:
Should we switch these two lines (same in the next check)? since `onChange` could potentially trigger the didUpdate synchronously?